cmd/daemon.go

```diff
@@ -126,22 +197,269 @@ func processPullRequestEvent(event *github.PullRequestEvent) {
+		return
+	}
+	defer cli.Close()
+
+	// Check for existing git repository, clone if no git repository exists.
+	err = checkForGit(projectDirectory)
+	if err != nil {
+		log.Println("No git repository present - cloning from POST event...")
+		auth, err := getGithubKey()
+		if err != nil {
+			http.Error(w, err.Error(), 500)
+			return
+		}
+
+		// Get github URL from up request for cloning
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			http.Error(w, err.Error(), 500)
+			return
+		}
+		defer r.Body.Close()
+		var upReq UpRequest
+		err = json.Unmarshal(body, &upReq)
+		if err != nil {
+			http.Error(w, err.Error(), 500)
+			return
+		}
+
+		// Clone project
+		remoteURL := upReq.Repo
+		log.Println("Attempting to clone " + remoteURL)
+		_, err = git.PlainClone(projectDirectory, false, &git.CloneOptions{
+			URL:  remoteURL,
+			Auth: auth,
+		})
+		if err != nil {
+			removeContents(projectDirectory)
+			http.Error(w, err.Error(), 500)
+			return
+		}
+	}
+
+	// Update and deploy the repository
+	repo, err := git.PlainOpen(projectDirectory)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	err = deploy(repo, cli)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	w.Header().Set("Content-Type", "text/html")
+	w.WriteHeader(http.StatusCreated)
+	fmt.Fprint(w, "Project up and running!")
 }
 
-// downHandler tries to bring the project down.
+// deploy does git pull, docker-compose build, docker-compose up
+func deploy(repo *git.Repository, cli *client.Client) error {
+	auth, err := getGithubKey()
+	if err != nil {
+		return err
+	}
+
+	// Pull from working branch
+	tree, err := repo.Worktree()
+	if err != nil {
+		return err
+	}
+	err = tree.Pull(&git.PullOptions{
+		Auth: auth,
+	})
+	if err != nil && err != git.NoErrAlreadyUpToDate {
+		// If pull fails, attempt a force pull before returning error
+		log.Println("Pull failed - attempting a fresh clone...")
+		repo, err = forcePull(repo, auth)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Kill active project containers if there are any
+	err = killActiveContainers(cli)
+	if err != nil {
+		return err
+	}
+
+	// Build and run project - the following code performs the
+	// shell equivalent of:
+	//
+	//    docker run -d \
+	// 	    -v /var/run/docker.sock:/var/run/docker.sock \
+	// 	    -v $HOME:/build \
+	// 	    -w="/build/project" \
+	// 	    docker/compose:1.18.0 up --build
+	//
+	// This runs a docker-compose image to build the Docker project.
+	// See https://cloud.google.com/community/tutorials/docker-compose-on-container-optimized-os
+	log.Println("Bringing project online.")
+	ctx := context.Background()
+	resp, err := cli.ContainerCreate(
+		ctx, &container.Config{
+			Image:      dockerCompose,
+			WorkingDir: "/build/project",
+			Env:        []string{"HOME:/build"},
+			Cmd:        []string{"up", "--build"},
+		},
+		&container.HostConfig{
+			Binds: []string{
+				"/var/run/docker.sock:/var/run/docker.sock",
```

Comment:
For mine and anyone else's notes - this is the second container that needs access to the docker socket. The daemon is bringing up a new container on the host (not docker-in-docker). This container has the sole intention of running `docker-compose up --build`.

Reply:
I'll add some more descriptive documentation in the daemon 👍